### ISSUE TITLE
Small question: where does the core of the method happen.

### ISSUE THREAD

**Issue by fraboeni** [Piyush-555/GaussianDistillation#2]
Thank you for providing the repo. It is a really interesting project. I am trying to understand where the core of the method happens. It seems that conceptually, the core is to update the BatchNorm statistics of the teacher when distilling with Gaussian noise: [here](https://github.com/Piyush-555/GaussianDistillation/blob/a5242dce0ff47c20b0592e8443940b3bb8209c91/distill_gaussian.py#L69). 

Is that already everything we need to consider, or is there something else, e.g. somewhere where we need to manually update the running statistics? I have not found anything like that in the repo.

Thanks and kind regards
Franziska

**Comment by Piyush-555** [comment]
Hi @fraboeni, thanks for your question!

Yes, that's it for the Gaussian distillation. So, putting the teacher model in training mode and not putting the student model in eval mode (missing line as it's already in training mode) during validation is all that is needed. Hence, both models will use BN stats based on the input batch. Running BN stats updates that are happening is just a side effect of how PyTorch handles things.

You have to manually update the running BN stats only in cases where you cannot perform inference on large enough batches (i.e., it's not possible to have good BN stats based on the input). This manual update is just forward passing your test data through the student model in training mode a bunch of times so that the running BN stats get updated and then you can use the student model in eval mode in whatever way you like.

Hope that helps. In case you have more questions, I'm keeping this issue open.
